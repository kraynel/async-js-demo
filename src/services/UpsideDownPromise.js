class UpsideDown {
  canBeKilled = false;

  enter = () => {
    console.info('Entering UpsideDown');
    let resolveRef = null;
    const promise = new Promise(resolve => {
      resolveRef = resolve;
    });
    setTimeout(resolveRef, 1000);
    return promise;
  };

  findDemogorgon = callback => {
    console.info('Looking for demogorgon');
    let resolveRef = null;
    const promise = new Promise(resolve => {
      resolveRef = resolve;
    });
    setTimeout(resolveRef, 1000, 'JUNKYARD');
    this.canBeKilled = true;
    return promise;
  };

  killDemogorgon = callback => {
    console.info('Trying to kill the demogorgon');
    let resolveRef = null;
    const promise = new Promise(resolve => {
      resolveRef = resolve;
    });
    setTimeout(() => {
      if (this.canBeKilled) {
        console.log('Demogorgon is dead!');
        resolveRef('SUCCESS');
        return;
      }
      console.log('You are dead!');
      resolveRef('DEAD');
    }, 1000);
    return promise;
  };
}

export default UpsideDown;
