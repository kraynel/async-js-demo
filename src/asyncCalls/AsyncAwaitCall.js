import React, { Component } from 'react';
import UpsideDownPromise from '../services/UpsideDownPromise';

export default class AsyncAwaitCall extends Component {
  constructor(props) {
    super(props);
    this.upsideDownPromise = new UpsideDownPromise();
    this.state = {
      strangerStatus: 'In real world'
    };
  }

  toggleCall = async () => {
    await this.upsideDownPromise.enter();
    this.setState({ strangerStatus: 'In upside down' });

    const location = await this.upsideDownPromise.findDemogorgon();
    this.setState({ strangerStatus: `Demogorgon in ${location}` });

    const result = await this.upsideDownPromise.killDemogorgon();

    if (result === 'SUCCESS') {
      this.setState({ strangerStatus: 'Demogorgon dead ️☠️' });
      return;
    }

    this.setState({ strangerStatus: 'You are dead ☠️' });
  };

  render() {
    return (
      <div>
        <button onClick={this.toggleCall}>Async/await Call</button>
        <span>{this.state.strangerStatus}</span>
      </div>
    );
  }
}
