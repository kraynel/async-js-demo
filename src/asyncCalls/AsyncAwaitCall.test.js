import React from 'react';
import renderer from 'react-test-renderer';

import UpsideDownPromise from '../services/UpsideDownPromise';
import AsyncAwaitCall from './AsyncAwaitCall';

jest.mock('../services/UpsideDownPromise');

describe('AsyncAwaitCall', () => {
  it('should start in the real world', () => {
    const component = renderer.create(<AsyncAwaitCall />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should go in the upside down', async () => {
    UpsideDownPromise.mockImplementation(() => {
      return {
        enter: () => Promise.resolve(),
        findDemogorgon: () => Promise.reject(new Error()),
        killDemogorgon: () => Promise.reject(new Error())
      };
    });

    const component = renderer.create(<AsyncAwaitCall />);
    try {
      await component.getInstance().toggleCall();
    } catch (error) {}
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should find the demogorgon', async () => {
    UpsideDownPromise.mockImplementation(() => {
      return {
        enter: () => Promise.resolve(),
        findDemogorgon: () => Promise.resolve('SOME PLACE'),
        killDemogorgon: () => Promise.reject(new Error())
      };
    });

    const component = renderer.create(<AsyncAwaitCall />);
    try {
      await component.getInstance().toggleCall();
    } catch (error) {}
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should kill the demogorgon', async () => {
    UpsideDownPromise.mockImplementation(() => {
      return {
        enter: () => Promise.resolve(),
        findDemogorgon: () => Promise.resolve('SOME PLACE'),
        killDemogorgon: () => Promise.resolve('SUCCESS')
      };
    });

    const component = renderer.create(<AsyncAwaitCall />);
    await component.getInstance().toggleCall();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should fail to kill the demogorgon', async () => {
    UpsideDownPromise.mockImplementation(() => {
      return {
        enter: () => Promise.resolve(),
        findDemogorgon: () => Promise.resolve('SOME PLACE'),
        killDemogorgon: () => Promise.resolve('FAILURE')
      };
    });

    const component = renderer.create(<AsyncAwaitCall />);
    await component.getInstance().toggleCall();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
