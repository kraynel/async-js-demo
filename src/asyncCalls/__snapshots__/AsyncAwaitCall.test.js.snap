// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncAwaitCall should fail to kill the demogorgon 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Async/await Call
  </button>
  <span>
    You are dead ☠️
  </span>
</div>
`;

exports[`AsyncAwaitCall should find the demogorgon 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Async/await Call
  </button>
  <span>
    Demogorgon in SOME PLACE
  </span>
</div>
`;

exports[`AsyncAwaitCall should go in the upside down 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Async/await Call
  </button>
  <span>
    In upside down
  </span>
</div>
`;

exports[`AsyncAwaitCall should kill the demogorgon 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Async/await Call
  </button>
  <span>
    Demogorgon dead ️☠️
  </span>
</div>
`;

exports[`AsyncAwaitCall should start in the real world 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Async/await Call
  </button>
  <span>
    In real world
  </span>
</div>
`;
